Drop support for numpy 1.x

mhvk opened the pull request:
This drops support for numpy < 2.0. It is in a few commits:
1. Oversight from a long time ago.
2. Drop numpy 1.25 (very few changes).
3. Drop numpy 1.26 (lots of changes).
4. Also remove `COPY_IF_NEEDED`, since it is now always just `None` (lots of changes).

EDIT: further commits
5. Bump other dependencies to the minimum needed for numpy >=2.0 support (inferred from https://github.com/numpy/numpy/issues/26191)
6. Silence a weird `RuntimeError` (happens on main too with same python/numpy combo; see #18989.

In total ~500 lines changed and ~600 lines deleted, i.e., lots of compatibility work-arounds simplified.

- [X] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

neutrinoceros reviewed (changes requested):
The situation might evolve but I don't think it is nearly motivated enough to make this change now, or in 6 months when we're looking at 8.0
Since this wasn't open as a draft I'm going to mark it as "changes requested" to prevent a merge, even though the real change I would request right now would be to drop the PR entirely.

nstarman reviewed (approved):
Approving for @astropy/cosmology 
I can review for units later.

I'm excited by the array api compatibility this will help enable!

nstarman commented:
I'm aware there is active discussion around timelines. My approval is for the code itself. Draft/request change reviews will ensure this PR isn't merged prematurely.

neutrinoceros commented:
> 1. Oversight from a long time ago.
> 1. Drop numpy 1.25 (very few changes).

btw both of these actually sound reasonable to me. I would approve them as an isolated PR

mhvk commented:
> Furthermore, @taldcroft brought up a real life example where astroppy not supporting numpy would create difficulties for his team. There's probably a way to keep supporting their needs _and_ drop support for numpy 1, but I think we should prioritize solving the problem _before_ we commit to a breaking change.

One can argue easily the other way: we need to do this early in the development cycle, as it brings out latent bugs, which we can fix before they cause real problems. I certainly *much* prefer to help fix issues like that than spend my time ensuring older versions keep working (for mostly mysterious users).

mhvk commented:
> this approach would only be practical if we retained the ability to revert this PR at any point. Because it's so big, I have doubts that this would hold; I'm worried that by merging this as an experiment, we de facto commit to it long term because we don't have the resources (man hours) for a clean revert. In other words: it's a trap !

I definitely don't intend this PR to be something to consider reverting. For the moment, it meant to show how much compatibility code we have, and therefore an argument to sticking to agreed policy (SPEC 0). If it gets merged, and there are not problems in the next few weeks or so, it will, as you note, become too difficult to revert. However, if there turn out to be problems later, *those affected by it* can either help fix the problems that numpy 2.0 poses for them (which I'd gladly assist in), choose to live with 7.2, or, if they need new features, start backporting them to 7.2 or so (which would be completely up to them as far as I'm concerned).

eerovaher commented on astropy/utils/compat/numpycompat.py:
```diff
@@ -21,12 +18,7 @@
 NUMPY_LT_2_4 = not minversion(np, "2.4.0.dev0")
-
-
-COPY_IF_NEEDED = False if NUMPY_LT_2_0 else None
```
https://github.com/astropy/astropy/blob/29217ffb870f5423b331a61206c6036d0016f640/astropy/utils/compat/__init__.py#L2-L5

neutrinoceros replied:
Nevermind I take this back (slightly): now would be the time to start enforcing it by *deprecating* the constant first (using a PEP 562 module-level `__getattr__`)

mhvk commented:
> > those affected by it can either help fix the problems that numpy 2.0 poses for them (which I'd gladly assist in)
> 
> The issue with the approach is that you're implicitly assuming that helping is within our reach. For anyone that relies on a locked system where only one version of numpy is available at a given time (be it some HPC server with highly restrictive maintenance rules, or a linux distro-level package manager), we wouldn't be able to provide any help, and by then, it's not a _choice_ for such users to "live under 7.2", but a necessity.

If they are so locked down that they are forced to live with numpy 1.*, then, yes, they will be forced to live with astropy 7.2 as well. Why should astropy be different from numpy? If they really need features from 8.0, they will have to do the work to backport them locally, or gang up together with others to create an astropy LTS version.

mhvk commented:
> As long as your branch has not diverged before the tag (and it should not), this should work.
> 
> ```
> git fetch upstream --tags
> git push origin v8.0.0.dev
> ```

OK, done that!
 
> > one test is the one @pllim suggested

This was just what you had mentioned, that when I pushed the tag, you could run the integration testing! I'm curious to see how bad it will get...

mhvk commented:
> Oh OK. I thought you meant like one single test somewhere. 😅

Thanks for running the tests. No horrible list of failures, thankfully! Just two failures, both of which seem not horrible (though of course one never knows).

pllim commented:
> Just two failures

I don't think those two are related. See https://github.com/astropy/astropy-integration-testing/pull/38#issuecomment-3578325673 :)

That said, this is just a first rough pass. Good that nothing broke but we should still do due diligence to communicate with downstream and give them time to test properly, if they have concerns.

neutrinoceros commented:
> it is definitely in a mode where we're trying to increase performance and compatibility.

my point exactly. Shedding off compatibility layers is a very clear case of going the opposite direction.

mhvk commented:
@pllim and @taldcroft - I agree with both of you. I think the only thing I would add is that any possible temporary LTS has to have a good reason, i.e., we discuss problems and possible work-arounds first, with LTS a clear last resort. I really mostly want to be sure that the burden of proof is on providing concrete problems, not hypothetical ones, and to try to find solutions within our standard APE framework. Also, if the problems do turn out to be so big that a "7.3" or "7.2.lts" is needed, one can and should expect those who really need it to help make it work (i.e., I really want to avoid moving work-for-me on keeping numpy 1.x compatibility alive to work-for-@pllim in supporting a LTS).

bsipocz commented:
> One way of doing that is to bump minimum versions as soon as possible in a development cycle. For downstream developers at least, that gives a proper chance to alert us.

☝️ This.

Independently of whether there will be an extended 7.2.x bugfix period or not, I would very strongly advocate for merging this asap once it's decided that the drop if going ahead for 8.0. The longer you let downstream to prep the better, this is not somehing one should realise in the RC period.

mhvk commented:
Now that we had agreement in the developer call to proceed (providing extended 7.2 support), I rebased and addressed the one comment so far (deprecating rather than removing `COPY_IF_NEEDED`).

mhvk commented:
@neutrinoceros - I'm fine with that, but hopefully you can help after to discover what the updated minimum requirements will be (it was quite a pain here, and I guess the list of packages for which to find out what version supports 2.x will now get longer...).

astrofrog commented on astropy/utils/compat/numpycompat.py:
```diff
@@ -21,12 +21,20 @@
 
-COPY_IF_NEEDED = False if NUMPY_LT_2_0 else None
+def __getattr__(attr):
+    # MHvK: Added in 8.0. Regular deprecation in 9.0, remove in 10.0?
```
Sounds good to me

mhvk commented:
@neutrinoceros - any idea what causes the failure on oldestdeps? This worked before the rebase, so I suspect a problem with versions, but just upgrading asdf-astropy to 0.8 does not seem to have been enough. Unfortunately, I cannot seem to test this locally - the doctests fail to work with that particular tox invocation...

pllim commented:
I think it is actually a bug in `main`. Why is `self` declared in a *function*? It is marked as "xfail" but maybe pytest is finally smart enough to smoke it out anyway?

https://github.com/astropy/astropy/blob/6156e3ad24f6c3ed34fac4ea9bfeaf3e2b0134c8/astropy/io/fits/tests/test_connect.py#L1038-L1040

pllim commented:
@braingram , any advise on which asdf-astropy should be the minversion if we bump to `numpy>=2`  here in `astropy`? Thanks!

pllim commented:
Yeah @neutrinoceros has been tightening the oldestdeps pinning since, so hopefully he can advise if braingram cannot.

braingram commented:
> @braingram , any advise on which asdf-astropy should be the minversion if we bump to `numpy>=2` here in `astropy`? Thanks!

It looks like the test environment picked up asdf 2.15.0:
https://github.com/astropy/astropy/actions/runs/20609735233/job/59192044395?pr=18986#step:10:38
that is too old for numpy 2.0. 2.15.1 might work as it includes a fix for the numpy scalar changes (https://github.com/asdf-format/asdf/issues/1605).

pllim commented:
As for linkcheck, it is unrelated but the URL https://ilrs.gsfc.nasa.gov/data_and_products/formats/cpf.html does appear to be broken. I opened https://github.com/astropy/astropy/issues/19112 separately.

braingram commented:
> Hmm, one failure is not obviously related, though it may be another case where we need to bump an indirect dependency: https://github.com/astropy/astropy/actions/runs/20622364534/job/59226665628?pr=18986#step:6:1920

Shows a 502 for a mast url. mast was (is?) in a bad state. I was told this should be fixed so rerunning this job might fix it.

neutrinoceros commented:
Indeed, we need at least 3.8.4: https://github.com/numpy/numpy/issues/26191

mhvk commented:
OK, I'll try following the guide and adjust...

neutrinoceros commented on astropy/io/fits/hdu/hdulist.py:
```diff
@@ -1528,24 +1527,12 @@ def _flush_resize(self):
-                            warnings.simplefilter("ignore", category=DeprecationWarning)
-                            arr.data = self[idx].data.data
-                    elif sys.getrefcount(arr) > 2:
+                    if sys.getrefcount(arr) > 2:
```
minor: we could now save one indentation level in what follows by using a single guard clause (`if mmap is None or sys.getrefcount(arr) <= 2`)

neutrinoceros commented on astropy/units/quantity_helper/function_helpers.py:
```diff
@@ -1325,7 +1264,7 @@ def isin(element, test_elements, *args, **kwargs):
 
 if NUMPY_LT_2_4:
-    # np.in1d deprecated in not NUMPY_LT_2_0, removed in not NUMPY_LT_24
+    # np.in1d removed in not NUMPY_LT_24
```
whoops
```suggestion
    # np.in1d removed in not NUMPY_LT_2_4
```

mhvk replied:
Rephrased as text, since it is now a simple if statement so no need for being greppable.

neutrinoceros commented on pyproject.toml:
```diff
@@ -83,17 +83,17 @@ all = [
     "mpmath>=1.2.1",
     "asdf-astropy>=0.7.0",
-    "bottleneck>=1.3.3",
+    "bottleneck>=1.4.0",  # minimum for numpy 2.0 compatibility
```
```suggestion
    "bottleneck>=1.4.0",
```

mhvk replied:
OK, those are indeed all in one commit, so easy enough to find if needed.

neutrinoceros commented on tox.ini:
```diff
@@ -60,19 +60,19 @@ description =
 
 deps =
-    numpy125: numpy==1.25.*
+    numpy20: numpy==2.0.*
```
I would actually suggest dropping this tox env now; I don't think it's used anywhere and it's not providing much value beyond `oldestdeps`

pllim replied:
I agree with this comment.

neutrinoceros reviewed (changes requested):
Here a complete review with a handful of nits, but overall nothing major to report.

neutrinoceros commented:
Before I reviewed I also pushed a commit to change `mpl380` to `mpl384`, but seeing that it makes for a more painful merge, I moved the commit to https://github.com/astropy/astropy/pull/19115

neutrinoceros commented:
That's already the content of #19115, and yes, it can wait for you :)

neutrinoceros commented:
That's already the content of #19115, and yes, it can wait for you :)

neutrinoceros commented:
Indeed. It'll create a couple trivial merge conflict but as long as it helps review, that's fine with me. Done.

mhvk commented on pyproject.toml:
```diff
@@ -43,8 +43,8 @@ keywords = [
-    "numpy>=1.25",
-    "pyerfa>=2.0.1.1",  # For >=2.0.1.7, adjust structured_units.rst and doctest-requires
+    "numpy>=2.0",
+    "pyerfa>=2.0.1.3",  # for numpy>=2.0; for >=2.0.1.7, adjust structured_units.rst and doctest-requires
```
Agreed, done

mhvk commented on pyproject.toml:
```diff
@@ -72,7 +72,7 @@ jupyter = [  # these are optional dependencies for utils.console and table
     # a direct dependency in order to set a lower bound for oldest-deps testing
     "jupyter-core>=4.11.2",
-    "pandas>=2.0.2",
+    "pandas>=2.2.2",  # minimum that supports numpy 2.0
```
OK, also done.

neutrinoceros reviewed (approved):
Thank you ! I trust you did go through all of them and I don't think I need a second full scan. Approving now !
